src/fluree/raft/log.clj

```diff
@@ -67,38 +76,169 @@
+                      e)))))
+
+(defn throw-corrupt-file-exception
+  "When reading a raft log file, report out a consistent format for exceptions if they occur."
+  [message exception ^RandomAccessFile raf file-name next-bytes index term entry-type entry-data]
+  (let [file-length  (.length raf)
+        file-pointer (.getFilePointer raf)
+        EOF?         (= file-length file-pointer)]
+    (throw (ex-info message
+                    {:status                 500
+                     :error                  :raft/corrupt-log
+                     :file-name              file-name
+                     :file-length            file-length
+                     :file-pointer           file-pointer
+                     :end-of-file-exception? EOF?
+                     :next-bytes             next-bytes
+                     :index                  index
+                     :term                   term
+                     :entry-type             entry-type
+                     :entry-data             entry-data}
+                    exception))))
+
+(defn read-next-entry
+  "Reads the next entry in a raft log file, returns 4-tuple of:
+  [index term entry-type entry-data] or an ex-info exception if there
+  was an error reading."
+  [^RandomAccessFile raf file-name]
+  (let [next-bytes (try (.readInt raf)
+                        (catch EOFException eof-ex
+                          (throw-corrupt-file-exception (str "Corrupt last entry in raft log: " file-name)
+                                                        eof-ex raf file-name nil nil nil nil nil))
+                        (catch Exception e (throw e)))
+        index      (try (.readLong raf)
+                        (catch EOFException eof-ex
+                          (throw-corrupt-file-exception (str "Corrupt last entry in raft log: " file-name)
+                                                        eof-ex raf file-name next-bytes nil nil nil nil))
+                        (catch Exception e (throw e)))
+        entry-type (if (pos? index) :append-entry (get entry-types' index))
+        term       (try (.readLong raf)
+                        (catch EOFException eof-ex
+                          (throw-corrupt-file-exception (str "Corrupt last entry in raft log: " file-name)
+                                                        eof-ex raf file-name next-bytes index nil entry-type nil))
+                        (catch Exception e (throw e)))
+        ba         (byte-array next-bytes)
+        _          (try (.read raf ba)
+                        (catch EOFException eof-ex
+                          (throw-corrupt-file-exception (str "Corrupt last entry in raft log: " file-name)
+                                                        eof-ex raf file-name next-bytes index term entry-type nil))
+                        (catch Exception e (throw e)))
+        entry-data (try (nippy/thaw ba)
+                        (catch Exception e
```

Comment:
could we catch a `java.io.EOFException` explicitly here along with a generic `Exception` to avoid the `EOF?` check below?

Reply:
No, because `(.read raf ba)` will not throw an exception if the byte array is larger than the remaining readable bytes so it won't throw an EOF exception. Let me think if there is a different approach however that can avoid the incorrect nippy deserialization I mentioned in the comment and be more fail-safe without impacting read performance too much.

Reply:
I was able to find the method `(.readFully raf ba)` instead of `.read`, which will throw an EOF exception now if the remaining bytes in the file are shorter than the byte array.

This change also ensures the nippy deserialization will not suffer the issue I mentioned in the last paragraph of the original PR. I'll update the PR comments.